### ISSUE TITLE
Synth graph optimization bug: missed case?

### ISSUE THREAD

**Issue by mtmccrea** [supercollider/supercollider#4382]
## Environment

* SuperCollider version: 3.10.2
* Operating system: macOS 10.14

## Steps to reproduce

```supercollider
(
s.waitForBoot({
	var func = {
		var out, src, foo;
		
		src = PinkNoise.ar;
		
		/* out is low-passed version of src */
		// out = LPF.ar(src);             // CASE 1: optimized out, as expected
		out = LPF.ar(src, 440, 4, 5);     // CASE 2: NOT optimized out.
		// foo = LPF.ar(src, 440, 4, 5);  // CASE 3: NOT optimized out.
		
		/* out is overwritten by src */
		out = src;
		Out.ar(3, out);
	};
	func.asSynthDef.children.do(_.postln);
})
)
```

## Expected vs. actual behavior
`out` is assigned to the `LPF`, but is then reassigned to `src`, making the `LPF` node in the graph obsolete; it should be optimized out.

`LPF` is properly omitted form the graph when using its "default" arguments (CASE 1). 
```supercollider
// SynthDef "children"
a Control
a PinkNoise
an Out
```
Specifying `mul` and/or `add` prevents the `LPF` from being optimized out (CASE 2). In this case, you'll see a `MulAdd` UGen added to the synth graph, which may be part of the reason the LPF isn't cleared from the graph.
```supercollider
// SynthDef "children"
a Control
a PinkNoise
a LPF
a MulAdd
an Out
```
Similarly, if you assign `LPF` to a variable that isn't used at all in the output (CASE 3), it still isn't cleared from the graph.
```supercollider
// SynthDef "children"
a Control
a PinkNoise
a LPF
a MulAdd
an Out
```

This is obviously a degenerate SynthDef, but I would thing the graph optimization should "see" this.

**Comment by mtmccrea** [comment]
FWIW, the case where this came up involved debugging a UGen that was doing in-place operation on a signal (the input/output buffers are aliased). 
I wanted to compare the output of the `src` alone, via plotting, vs. that signal filtered by the UGen under test. I noticed that when sending `src` out directly (assigning it to `out` downstream from the UGen under test), the UGen was still effecting the signal, though it wasn't explicitly used in the output. So the UGen was still doing work, processing the signal, though shouldn't have been a causal part of the output (I would think).

**Comment by jamshark70** [comment]
After a quick look -- `PureUGens` can do dead code elimination; others cannot. (Dead code elimination is ok for pure UGens because a pure UGen has no side effects -- its only effect is its output, and if the output isn't plugged into anything, then the UGen wouldn't be missed.)

Math operators haven't been marked as PureUGens. But, unary and binary ops do call `performDeadCodeElimination`. `MulAdd` misses this step. I think there's no harm to add it.

But...

> the case where this came up involved debugging a UGen that was doing in-place operation on a signal (the input/output buffers are aliased).

If the UGen you're debugging has a side effect, then it can't be a pure UGen and it would not be a candidate for dead code elimination. So, while you're correct about MulAdd, it wouldn't help your original case.

**Comment by snappizz** [comment]
my apologies if you already know this, but please don't do `mul` and `add` arguments in new ugens. they were originally there for efficiency but they are now a legacy feature since SC optimizes `+` and `*` to a `MulAdd`. they also put you in an awkward position when you want to add new arguments later but retain backward compatibility.

**Comment by jamshark70** [comment]
Related: `Sum3` and `Sum4`:

```
(
var func = {
	var out, src, foo;
	
	src = PinkNoise.ar;
	out = Sum3(src, src, src);
	
	/* out is overwritten by src */
	out = src;
	Out.ar(3, out);
};
func.asSynthDef.children.do(_.postln);
)

a Control
a PinkNoise
a Sum3
an Out
```

Changing MulAdd, Sum3 and Sum4 to inherit from PureUGen fixes the issue.

**Comment by mtmccrea** [comment]
Thanks @jamshark70 for the tip on `PureUGen`s. I didn't know there were UGens excluded from dead code elimination, but after looking into what distinguishes a `PureUGen` from others, this makes sense. 

> But...
> 
> >the case where this came up involved debugging a UGen that was doing in-place operation on a signal (the input/output buffers are aliased).

> If the UGen you're debugging has a side effect, then it can't be a pure UGen and it would not be a candidate for dead code elimination. 
> 

I'm not sure that in-place operation is a "side effect". Many UGens do it, and it's a good design for efficient resource management and non-trivial speedups in the case of multichannel UGens, so it would be unfortunate if this disqualified it from optimization. It doesn't appear to be the case—after just switching my UGen to inherit from `PureMultiOutUGen`, the "bug" disappeared :) I'll put it on my TODO to see that `PureUGen`s are mentioned (and encouraged) in the docs on writing UGens.

And thanks @snappizz for the reminder about `mul` and `add`. I'm not using them, I was just trying to create a minimal reproducer and I got lucky in choosing `LPF`, with arguments provided, and the culprit happened to be `MulAdd`.

In terms of actionable follow-up—is there a reason `MulAdd` can't inherit from `PureUgen`? Similarly `Sum3` and `Sum4`, as James mentioned. 

A further step would be a review of all of the UGens to see which others may have been overlooked? In the case of the latter, is this filed Issue enough to "queue it up"? I could change the name of this Issue if that's helpful: "Review UGens for PureUGen inheritance". Thoughts?

or is there another place to log larger codebase review tasks? [long-term backlog](https://github.com/supercollider/supercollider/projects/2)?

Tangent: when making these kinds of changes, how confident are you with the current Unit Tests to ensure there aren't unforeseen consequences? Within reason...

**Comment by jamshark70** [comment]
Do they? Maybe I'm missing something, but that sounds like a problem.

"the UGen was still effecting the signal, though it wasn't explicitly used in the output"

That's... not OK. That's not OK at all. If we have core UGens that are doing this, that's very, very bad.

A UGen's output can plug into multiple inputs of other units.

```supercollider
(
SynthDef(\test, {
	var a = SinOsc.ar,
	b = LPF.ar(a),
	c = a + b;
	Out.ar(0, c);
}).draw;
)
```

**

The `+` operator *must* receive an unmodified `a` signal and an unmodified `b` signal. If `b` overwrites `a`'s output wire buffer, then its result is invalid.

What you're describing is totally unacceptable. Any core UGens that are doing this need to be fixed. "... it's a good design for efficient resource management and non-trivial speedups in the case of multichannel UGens" but if it means that downstream units receive incorrect inputs, then it's an invalid optimization.

Now, in some 17-18 years of using SC, I've never heard of any core UGens doing this. So I suppose this is more likely to be a misunderstanding than an actual bug.

> In terms of actionable follow-up—is there a reason MulAdd can't inherit from PureUgen? Similarly Sum3 and Sum4, as James mentioned.

The risk here is basically 0. MulAdd combines `+` and `*`, both of which do dead code elimination. Sum3 and Sum4, `+`, same.

> A further step would be a review of all of the UGens to see which others may have been overlooked? In the case of the latter, is this filed Issue enough to "queue it up"?

No need. https://github.com/supercollider/supercollider/issues/3371

**Comment by mtmccrea** [comment]
> So I suppose this is more likely to be a misunderstanding than an actual bug.

Hmmm, I'll have to think harder on this then.  Your example makes it clear why this would be a problem.  

I have been working under the assumption that buffer aliasing means that input and output buffers are actually the same memory location. It's a common convention in UGens that the input at index _x_ must be read before writing the output at index _x_ for this reason. Sticking with `LPF` as an example, a quick insertion into the source, seems to validate this:

LPF source:
```c++
void LPF_next(LPF* unit, int inNumSamples)
{
	printf("[LPF] in  pointer: %p\n", ZIN(0));
	printf("[LPF] out pointer: %p\n", ZOUT(0));

...
PluginLoad(Filter)
{
	ft = inTable;
	...
	DefineSimpleUnit(LPF); // << LPF is allowed to alias its i/o buffers
```
Output in SC:
```
[LPF] in  pointer: 0x7faf0c0f5bfc
[LPF] out pointer: 0x7faf0c0f5bfc // i/o pointers point to the same location: i/o buffers are identical
```

I did a similar test in the case of the UGen I'm writing which uses the `SCUnit` struct, so I/O buffer pointers are accessed with `mInBuf` and `mOutBuf`, and I get the same result:
```
inBuf pointers
	mInBuf[0]:	0x7f80450bd700
	mInBuf[1]:	0x7f80450bd800
	mInBuf[2]:	0x7f80450bd900
	mInBuf[3]:	0x7f80450bda00
outBuf pointers
	mOutBuf[0]:	0x7f80450bd700
	mOutBuf[1]:	0x7f80450bd800
	mOutBuf[2]:	0x7f80450bd900
	mOutBuf[3]:	0x7f80450bda00
```

So I'm missing something important here. I'll keep digging to figure out the difference...

And FWIW, from the SuperCollider book, Ch. 25 Writing Unit Generator Plug-ins:
> The server uses a “buffer coloring” algorithm to minimize use of buffers and to optimize cache performance. This means that any of the output buffers may be the same as 1 of the input buffers. This allows for in-place operation, which is very efficient. You must be careful, however, not to write any output sample before you have read the corresponding input sample. If you break this rule, then the input may be overwritten with output, leading to undesired behavior. If you can’t write the UGen efficiently without breaking this rule, then you can instruct the server not to alias the buffers by using the `DefineSimpleCantAliasUnit()`􏰆􏰋 or `DefineDtorCantAliasUnit()` macros in the `load()` function, rather than the `DefineSimpleUnit()`􏰆􏰋 or `DefineDtorUnit()`􏰋 macros.

**Comment by jamshark70** [comment]
The answer (which I expected, and just confirmed) is that SC's wire buffer allocation strategy accounts for outputs connected to multiple inputs.

I haven't looked at that part of the code at all, but, moving the 'printf' statements into SC_Graph.cpp:

```
(
fork {
	d = SynthDef(\test, {
		var a = DC.ar(1),
		b = HPF.ar(a, 1000),
		c = a + b;
		Out.ar(1000, c);  // DC->speakers = bad
	}).add;
	
	d.children.do { |unit, i| "%: %\n".postf(i, unit) };
	
	s.sync;
	a = Synth(\test);
};
)

0: a DC
1: a HPF
2: a BinaryOpUGen
3: an Out
unit 0 inbuf 0 = 0x7f8647dd0248
unit 0 outbuf(ar) 0 = 0x560452892340
unit 1 inbuf 0 = 0x560452892340  // HPF audio input === DC output
unit 1 inbuf 1 = 0x7f8647dd0268
unit 1 outbuf(ar) 0 = 0x560452892440  // HPF output != its input
unit 2 inbuf 0 = 0x560452892340  // + input == DC
unit 2 inbuf 1 = 0x560452892440  // + input == HPF
unit 2 outbuf(ar) 0 = 0x560452892340  // + output reuses DC wirebuf
unit 3 inbuf 0 = 0x7f8647dd0268
unit 3 inbuf 1 = 0x560452892340
```

So somewhere in SC's code base, it's recognizing that DC.ar output will be needed later and HPF should not overwrite that wire buffer.

OK, let's try the original case. First, DC --> HPF --> MulAdd --> Out (all units used). It does optimize wire buffer usage.

```
(
fork {
	d = SynthDef(\test, {
		var a = DC.ar(1),
		b = HPF.ar(a, 1000, 0.5, 0.5);
		Out.ar(1000, b);  // DC->speakers = bad
	}).add;
	
	d.children.do { |unit, i| "%: %\n".postf(i, unit) };
	
	s.sync;
	a = Synth(\test);
};
)

0: a DC
1: a HPF
2: a MulAdd
3: an Out
unit 0 inbuf 0 = 0x7f8647dd0248
unit 0 outbuf(ar) 0 = 0x560452892340
unit 1 inbuf 0 = 0x560452892340
unit 1 inbuf 1 = 0x7f8647dd0268
unit 1 outbuf(ar) 0 = 0x560452892340  // filter output === input
unit 2 inbuf 0 = 0x560452892340
unit 2 inbuf 1 = 0x7f8647dd0288
unit 2 inbuf 2 = 0x7f8647dd0288
unit 2 outbuf(ar) 0 = 0x560452892340
unit 3 inbuf 0 = 0x7f8647dd0268
unit 3 inbuf 1 = 0x560452892340
```

If I change to `Out.ar(1000, a)` (bypassing `b`), then:

```
0: a DC
1: a HPF
2: a MulAdd
3: an Out
unit 0 inbuf 0 = 0x7fa60e96c248
unit 0 outbuf(ar) 0 = 0x562cb9fd9540
unit 1 inbuf 0 = 0x562cb9fd9540
unit 1 inbuf 1 = 0x7fa60e96c268
unit 1 outbuf(ar) 0 = 0x562cb9fd9640  // filter output != input
unit 2 inbuf 0 = 0x562cb9fd9640
unit 2 inbuf 1 = 0x7fa60e96c288
unit 2 inbuf 2 = 0x7fa60e96c288
unit 2 outbuf(ar) 0 = 0x562cb9fd9640
unit 3 inbuf 0 = 0x7fa60e96c268
unit 3 inbuf 1 = 0x562cb9fd9540  // Out looks to DC's input
```

All cases, with core units, are handled correctly.

Therefore, the UGen where you got the stray output effect must be doing something funky with the inputs and outputs that is breaking the wire buffer assignment logic. (But that's about as far as I can go... I haven't yet found the part of the code that determines whether a wire may be reused or not.)

Oh, my pointer-printing patch:

```
--- a/server/scsynth/SC_Graph.cpp
+++ b/server/scsynth/SC_Graph.cpp
@@ -392,6 +392,7 @@ void Graph_Ctor(World *inWorld, GraphDef *inGraphDef, Graph *graph, sc_msg_iter
                                Wire *wire = graphWires + inputSpec->mWireIndex;
                                unitInput[j] = wire;
                                unitInBuf[j] = wire->mBuffer;
+                               printf("unit %d inbuf %d = %p\n", i, j, unitInBuf[j]);
                        }
                }
 
@@ -412,6 +413,7 @@ void Graph_Ctor(World *inWorld, GraphDef *inGraphDef, Graph *graph, sc_msg_iter
                                        wire->mBuffer = bufspace + outputSpec->mBufferIndex;
                                        unitOutput[j] = wire;
                                        unitOutBuf[j] = wire->mBuffer;
+                                       printf("unit %d outbuf(ar) %d = %p\n", i, j, unitOutBuf[j]);
                                }
                                calcUnits[calcCtr++] = unit;
                        } else {
@@ -421,6 +423,7 @@ void Graph_Ctor(World *inWorld, GraphDef *inGraphDef, Graph *graph, sc_msg_iter
                                        wire->mBuffer = &wire->mScalarValue;
                                        unitOutput[j] = wire;
                                        unitOutBuf[j] = wire->mBuffer;
+                                       printf("unit %d outbuf(kr) %d = %p\n", i, j, unitOutBuf[j]);
                                }
                                if (unitCalcRate == calc_BufRate) {
                                        calcUnits[calcCtr++] = unit;
```

**Comment by mtmccrea** [comment]
> The answer (which I expected, and just confirmed) is that SC's wire buffer allocation strategy accounts for outputs connected to multiple inputs.

This is a crucial insight, thanks for the deep dive on this, James. So during allocation, by a strategy outside our control when composing a UGen, `mInBuf` and `mOutBuf` **may or may not** be equal, even when buffer aliasing is permitted. And in fact after some tweaking of the graph in my `LPF` example, I was able to get them to not be equal.

This is particularly helpful in my case, I now know that my initial assumption was wrong. It led me to believe that when explicitly allowing buffer aliasing I could simply use either of the `mInBuf` and `mOutBuf` pointers for both reading input and writing output. I'll be curious if now having to use both pointers will slow things down a bit... 

So IOW the efficiency gains from in-place operations is a nice idea in principle, and you should try to use it when you can, but you're not *guaranteed* to see the benefit from this if the wire buf allocator decides you can't use a shared buffer.

So the description from the SC book

> This means that any of the output buffers ***may*** be the same as 1 of the input buffers. [emphasis mine]

...is strictly speaking true, but not the full story.

So given that #3371 encompasses the actionable takeaway from this issue, this issue can probably be closed.

**Comment by jamshark70** [comment]
> I'll be curious if now having to use both pointers will slow things down a bit...
> So IOW the efficiency gains from in-place operations is a nice idea in principle, and you should try to use it when you can, but you're not guaranteed to see the benefit from this if the wire buf allocator decides you can't use a shared buffer.

I think the baseline for performance should not be considered to be the aliased case (same wire buffer for input and output). The baseline is separate wire buffers, because that is certain to get the right result in every case.

Collapsing an output wire buffer to reuse one of its input wire buffers is an optional optimization, which applies only if the input has one and only one output.

Code shouldn't assume that the optimized case is normal. It isn't (though it may be reasonably common). Normal is un-optimized. The optimization is nice to have when it applies.

Glad that we seem to have a satisfactory answer -- things are working as they are, and you have more information about writing your UGen.

However, I do want to keep this open because we should fix MulAdd, Sum3 and Sum4 -- those are pretty simple cases that don't need to wait for a comprehensive review of PureUGens. I'll take care of that, once codebase reformatting is complete.

**Comment by mtmccrea** [comment]
> Collapsing an output wire buffer to reuse one of its input wire buffers is an optional optimization...

Right. The takeaway for me is that when we write UGens, this buffer aliasing optimization is not actually within our control. The only "optional" control we have is preventing it outright with the `kUnitDef_CantAliasInputsToOutputs` flag. We can't (or shouldn't) expect this optimization to happen, and should address input and output buffer pointers separately.

> ...which applies only if the input has one and only one output.

This might need some clarification, as you'll see in my multichannel I/O [example](https://github.com/supercollider/supercollider/issues/4382#issuecomment-482893538) above, it also appears to work on multi-in multi-out UGens. Without seeing exactly where this optimization takes place, I would imagine intuitively that the constraint is: buffer aliasing is only available when number of outputs is equal to (or possibly also less than?) the number of inputs.

>  I do want to keep this open because we should fix MulAdd, Sum3 and Sum4 ... I'll take care of that, once codebase reformatting is complete.

Great thanks!

**Comment by mtmccrea** [comment]
Ugh... some initial tests show a significant slowdown when you aren't able to do in-place operation (in my case in-place processing is ~1.6x faster). Even when input and output buffers are aligned, by simply having to read from one, set the other, increment/do math on multiple pointers. My case has high channel counts (dozens of channels), so maybe there's a cumulative effect as channel count increases. 

One solution is to write multiple calc functions, check for I/o buffer pointer alignment in the ctor, and call the in-place operation calc function if the pointers align.  If the pointers align, theoretically the the graph builder has already determined this is "safe", and the dead code elimination issue we've been discussing should be mitigated by inheriting from `PureMultiOutUGen`.

In any case, this is now a bit OT for this issue, maybe I'll start a new thread elsewhere if I have more definitive insights or concise cases to discuss.

**Comment by jamshark70** [comment]
> ...which applies only if the input has one and only one output.

Hm, I don't know what I was thinking to try to explain it like that. I should have said "only if the source output channel is connected to one and only one target UGen's input" (if it's connected to multiple inputs or multiple UGens, then the original signal must be preserved).

It's unfortunate that an invalid optimization got your hopes up. I don't have any suggestions to improve the performance.
